modules/core/src/test/java/org/apache/ignite/internal/processors/compute/IgniteComputeWithPendingTopologyDeadlockTest.java

```diff
@@ -0,0 +1,313 @@
+import org.apache.ignite.cache.CacheMode;
+import org.apache.ignite.cache.CacheWriteSynchronizationMode;
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.IgniteInterruptedCheckedException;
+import org.apache.ignite.internal.util.typedef.PA;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionIsolation;
+
+import static org.apache.ignite.testframework.GridTestUtils.assertThrows;
+import static org.apache.ignite.testframework.GridTestUtils.runAsync;
+
+/**
+ * Test that compute task will be rejected when executed within transaction/lock if topology pending updates.
+ */
+public class IgniteComputeWithPendingTopologyDeadlockTest extends GridCommonAbstractTest {
+
+    /** Maximum timeout for topology update (prevent hanging and long shutdown). */
+    private static final long PENDING_TIMEOUT = 15_000L;
+
+    /** */
+    private static final long CACHE_CREATION_TIMEOUT = 2_000L;
+
+    /** */
+    private static final int CACHE_SIZE = 10;
+
+    /** */
+    private static final String CACHE1 = "cache1";
+
+    /** */
+    private static final String CACHE2 = "cache2";
+
+    /** */
+    private static final String CACHE3 = "cache3";
+
+    /** */
+    private static final String EXPECTED_ERROR_MESSAGE = "Pending " +
+        "topology found - job execution within transaction or lock was canceled.";
+
+    /** {@inheritDoc} */
+    @Override public void beforeTest() throws Exception {
+        super.beforeTest();
+        startGrid(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void afterTest() throws Exception {
+        super.afterTest();
+        stopAllGrids();
+    }
+
+    /**
+     * Create cache configuration with transactional atomicity mode.
+     *
+     * @param name Cache name.
+     * @param mode {@link CacheMode Cache mode.}
+     * @param writeMode {@link CacheWriteSynchronizationMode Cache write synchronization mode.}
+     * @return Cache configuration.
+     */
+    private CacheConfiguration<Integer, Integer> cacheConfiguration(String name, CacheMode mode,
+        CacheWriteSynchronizationMode writeMode) {
+        CacheConfiguration<Integer, Integer> cacheConfiguration = new CacheConfiguration<>(name);
+
+        cacheConfiguration.setCacheMode(mode);
+        cacheConfiguration.setWriteSynchronizationMode(writeMode);
+        cacheConfiguration.setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL);
+
+        return cacheConfiguration;
+    }
+
+    /**
+     * Explicit lock test.
+     *
+     * @throws Exception If fails.
+     */
+    public void testExplicitLock() throws Exception {
+        for (final CacheMode mode : CacheMode.values()) {

```

Comment:
Please, make 3 tests instead this for loop

`testExplicitLockLocal`
`testExplicitLockReplicated`
`testExplicitLockPartitioned`

In case of any future bugs it speed up understanding and fixing.